### PR TITLE
Test/book

### PR DESCRIPTION
**Pull request by wave49192**
Implemented
- add gem for testing
- book model spec file
- book request spec file
- config testing gem in rails_helper.rb

### CODE REVIEW

File: spec/models/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe Book, type: :model do
+  describe 'valid factory' do
```
Comment (siripoom-datawow) [inline spec/models/book_spec.rb:4]: For describe message please start with `#` or `.`
Example `describe '#valid factory' do` or  `describe '.valid factory' do`

File: spec/models/book_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  describe 'release attribute' do
+    subject(:book) { build(:book) }
+    it 'allows valid dates' do
+      expect(book).to be_valid
+    end
+
+    it 'handles nil release date' do
+      expect(book).to be_valid
+    end
```
Comment (siripoom-datawow) [inline spec/models/book_spec.rb:48]: what different between `'release attribute'` and `'handles nil release date'`
both of it test the same case with same condition?
recheck again krub

File: spec/requests/books_spec.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        patch book_path(book), params: { book: { name: new_name } }
+        book.reload
+        expect(book.name).to eq(new_name)
+        expect(response).to redirect_to(book_path(book))
+      end
+    end
+
+    context "with invalid parameters" do
+      it "does not update the book and redirects to the edit page" do
+        patch book_path(book), params: { book: { name: nil } }
```
Comment (siripoom-datawow) [inline spec/requests/books_spec.rb:83]: FYI for `params` you can declare `patch book_path(book), params` on the top after `context`. If there are many `it`, you just need to to re-assign params and use it
BTW Your choice is good, i really love it.

File: spec/models/book_spec.rb (hunk 1 above)
Comment (wave49192) [reply spec/models/book_spec.rb:4]: From what I discussed with P'Min, The `#` and `.` prefixes are for the instances method and class method respectively. Not for the Model or requests itself.

File: spec/models/book_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+require 'rails_helper'
+
+RSpec.describe Book, type: :model do
+  describe 'valid factory' do
+    subject(:factory) { build(:book) }
+    it { is_expected.to be_valid }
```
Comment (WatcharitNew) [inline spec/models/book_spec.rb:6]: May add space to separate subject and it block

File: spec/requests/books_spec.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      expect(response).to be_successful
+
+      books.each do |book|
+        expect(response.body).to include(book.name)
+      end
+    end
+  end
+
+  describe "GET /show" do
+  let!(:book) { books.first }
```
Comment (WatcharitNew) [inline spec/requests/books_spec.rb:23]: indent issue. Make it align with describe block

File: spec/requests/books_spec.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  describe "GET /new" do
+    it "returns a successful response" do
+      get new_book_path
+      expect(response).to be_successful
+    end
+  end
+
+  describe "GET /edit" do
+  let!(:book) { books.first }
```
Comment (WatcharitNew) [inline spec/requests/books_spec.rb:40]: this too

File: spec/requests/books_spec.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        invalid_params = attributes_for(:book, name: nil)
+        post books_path, params: { book: invalid_params }
+        expect(response).to redirect_to(new_book_path(error: [ "Name can't be blank" ]))
+        expect(Book.count).to eq(books.size)
+      end
+    end
+  end
+
+  describe "PATCH /update" do
+     let(:book) { books.first }
```
Comment (WatcharitNew) [inline spec/requests/books_spec.rb:69]: this one as well

**Pull request merged**